[nikohomecontrol] Alarm systems

mherwege opened the pull request:
Add support for alarm systems in Niko Home Control

mherwege commented:
@lolodomo This code was building on the previous PR https://github.com/openhab/openhab-addons/pull/12893 you reviewed and has been used by some users for more then a year. It adds another device type to the code. It was a bit of a challenge to get this re-integrated in the code.

I hope this is quicker to review, as the structure is exactly the same as for the other devices that have been added in the previous PR. I have gone through the review feedback of the previous PR and applied these to this PR. I hope I didn't miss too many.

I would be very grateful if you could review this PR as well. That would bring the code in the distribution up to the level I have locally and would be a good basis to build in the future.

lolodomo commented:
I have still in mind the previous PR and I will review this one soon.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -508,6 +528,42 @@
+		<label>@text/channelAlarmArmedLabel</label>
+		<description>@text/channelAlarmArmedDescription</description>
+		<category>Alarm</category>
+		<autoUpdatePolicy>veto</autoUpdatePolicy>
```
For my curiosity, why setting this policy ?

mherwege replied:
When you arm, it will take e.g. 30 seconds before it is fully armed, typically the time to leave the home. The armed state should therefore not be set directly by the UI, but only after 30s when there is a feedback from the alarm system it is effectively armed. The veto policy makes sure to only set based on what comes back from the device, not use the initial UI set value.

lolodomo replied:
Makes sense.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlAlarmHandler.java:
```diff
@@ -0,0 +1,217 @@
+    private void updateProperties(NhcAlarm nhcAlarm) {
+        Map<String, String> properties = new HashMap<>();
+
+        if (nhcAlarm instanceof NhcAlarm2) {
+            NhcAlarm2 alarm = (NhcAlarm2) nhcAlarm;
```
Only suggestion: nhcAlarm instanceof NhcAlarm2 alarm

mherwege replied:
Done

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlAlarmHandler.java:
```diff
@@ -0,0 +1,217 @@
+            return;
+        }
+
+        updateState(CHANNEL_CONTROL,
+                NHCOFF.equals(state) || NHCDETECTORPROBLEM.equals(state) ? OnOffType.OFF : OnOffType.ON);
+        updateState(CHANNEL_ARMED,
+                NHCOFF.equals(state) || NHCPREARMED.equals(state) || NHCDETECTORPROBLEM.equals(state) ? OnOffType.OFF
+                        : OnOffType.ON);
```
Only suggestion: use OnOffType.from

mherwege replied:
I don't see how as some of these don't match on/off or true/false.

lolodomo replied:
`My idea: updateState(CHANNEL_CONTROL, OnOffType.from(NHCOFF.equals(state) || NHCDETECTORPROBLEM.equals(state)));`

mherwege replied:
That’s a possibility. This is probably personal, but for me that does not look clearer. So I am not inclined to change it.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlAlarmHandler.java:
```diff
@@ -0,0 +1,217 @@
+        }
+
+        if ((CHANNEL_CONTROL.equals(channelUID.getId()) || CHANNEL_ARMED.equals(channelUID.getId()))
+                && command instanceof OnOffType) {
+            OnOffType s = (OnOffType) command;
```
Only suggestion: command instanceof OnOffType s

mherwege replied:
Done

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/handler/NikoHomeControlAlarmHandler.java:
```diff
@@ -0,0 +1,217 @@
+        }
+
+        triggerChannel(CHANNEL_ALARM);
+        updateStatus(ThingStatus.ONLINE);
```
I believe we had this discussion about setting thing status when triggering a channel. I do not remember the conclusion.
So just check that it is consistent.

mherwege replied:
Indeed, and I kept it, as this is coming from the NHC system, and receiving something effectively means the handler is online, even when it didn't report back before.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/README.md:
```diff
@@ -206,10 +213,12 @@ Thing nikohomecontrol:accessRingAndComeIn:mybridge:myaccess [ accessId="abcdef01
-The event message is the alarm or notice text coming from Niko Home Control.
+| control         | RW |          | Switch             | alarm       | arm/disarm alarm                                                                                     |
+| armed           | RW |          | Switch             | alarm       | state of the alarm system (on/off), will only turn on after pre-armed period when arming             |
+| targetstate     | RW |          | Switch             | alarm       | target state of the alarm system (ARMED/DISARMED), will change state immediately                     |
```
Type is String

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -508,6 +528,42 @@
+		<category>Alarm</category>
+		<autoUpdatePolicy>veto</autoUpdatePolicy>
+	</channel-type>
+	<channel-type id="alarmtargetstate">
+		<item-type>String</item-type>
+		<label>@text/channelAlarmTargetStateLabel</label>
+		<description>@text/channelAlarmTargetStateDescription</description>
+		<category>Switch</category>
+		<command>
+			<options>
+				<option value="DISARMED">@text/channelOptionAlarmStateDisarmed</option>
+				<option value="ARMED">@text/channelOptionAlarmStateArmed</option>
+			</options>
+		</command>
+	</channel-type>
```
I have a doubt, shouldn't it be rather a Switch ?

mherwege replied:
That would be the control switch that is already there. In that case I wouldn't need this channel. This channel was explicitly added for compatibility with both Homekit and Google Assistent. I believe it was Google Assistent that required this to be a String. Things may have changed since, as I have seen there has been quite a bit of change on Google Assitent side.
See discussion here: https://community.openhab.org/t/niko-home-control-ii/44586/453?u=mherwege

mherwege replied:
I requested confirmation: https://community.openhab.org/t/niko-home-control-ii/44586/468?u=mherwege

lolodomo replied:
Rather than "duplicating" channel, are you sure that the user could not simply create a String item linked to the Switch channel and set command options ?

That is a little strange if now Homekit and Google Assistent adds constraints to the design of a binding. especially when this is something not documented at my knowledge.

mherwege replied:
For the Homekit security system, the required item types are Number or String (https://next.openhab.org/addons/integrations/homekit/). For Google Assistant, the controlling item needs to be a Switch (https://next.openhab.org/docs/ecosystem/google-assistant/#securitysystem-as-switch). I use neither of these, so cannot test if connecting a String item to a switch channel with mapping options would work. I would prefer being able to limit the number of channels and to me a Swich channel would indeed make most sense.

mherwege replied:
For the Homekit security system, the required item types are Number or String (https://next.openhab.org/addons/integrations/homekit/). For Google Assistant, the controlling item needs to be a Switch (https://next.openhab.org/docs/ecosystem/google-assistant/#securitysystem-as-switch). I use neither of these, so cannot test if connecting a String item to a switch channel with mapping options would work. I would prefer being able to limit the number of channels and to me a Swich channel would indeed make most sense.

mherwege replied:
> Rather than "duplicating" channel, are you sure that the user could not simply create a String item linked to the Switch channel and set command options ?

@lolodomo I now tried this out. Unless I missed something, it doesn't work. You cannot change type through a profile and a Switch channel does not work with a String item. Even if a profile would be available, it is only one way, from channel to item. So the only way to achieve it with one channel is by using a rule that makes the translation.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlCommunication.java:
```diff
@@ -192,6 +193,15 @@ public Map<String, NhcVideo> getVideoDevices() {
+    /**
+     * Return all alarm devices in the Niko Home Control Controller.
+     *
+     * @return <code>Map&lt;String, {@link NhcAlarm}></code>
```
> will only need escaping when in a pair with <. As < is escaped, that's not the case, so fine.

mherwege commented on bundles/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NhcAlarm.java:
```diff
@@ -0,0 +1,176 @@
+    }
+
+    /**
+     * This method should be called when an object implementing the {@NhcAccessEvent} interface is disposed.
```
Done

mherwege commented:
@lolodomo As there is no feedback so far, I simplified the code and removed the `targetState` String channel. There is only one Switch channel now to control the alarm system (I renamed it to `arm`). There is some extra text in the README to explain. I hope that we we can progress this.

lolodomo commented on bundles/org.openhab.binding.nikohomecontrol/README.md:
```diff
@@ -265,6 +275,9 @@ String ThermostatDemand   {channel="nikohomecontrol:thermostat:nhc1:5:heatingdem
 Number:Power CurPower   "[%.0f W]"  {channel="nikohomecontrol:energyMeterLive:nhc1:6:power"} # Get current power consumption
 Number:Energy MyMeter          "[%.0f kWh]"   {channel="nikohomecontrol:energyMeter:nhc1:7:energy         # Get energy meter reading
 Number:Energy MyMeterDay       "[%.0f kWh]"   {channel="nikohomecontrol:energyMeter:nhc1:7:energyday      # Get energy meter day reading
+Switch Alarm            {channel="nikohomecontrol:onOff:nhc2:7:targetstate"}       # Switch to arm/disarm alarm with option values (ARMED/DISARMED)
```
Please fix channel

lolodomo reviewed (approved):
LGTM, thank you
